### PR TITLE
SI-6863 root cause fixed using factory of scala.runtime.*Ref

### PR DESCRIPTION
**Pull request by magarciaEPFL**
This PR supercedes https://github.com/scala/scala/pull/2460

review by @JamesIry or @retronym

### CODE REVIEW

File: src/compiler/scala/tools/nsc/transform/LambdaLift.scala [hunk 1]
```diff
@@ -444,56 +459,21 @@ abstract class LambdaLift extends InfoTransform {
...
             val tpt1 = TypeTree(sym.tpe) setPos tpt.pos
-            /* Creating a constructor argument if one isn't present. */
-            val constructorArg = rhs match {
+
+            val refTypeSym = sym.tpe.typeSymbol
+
+            val factoryCall = typer.typedPos(rhs.pos) {
+              rhs match {
               case EmptyTree =>
-                sym.tpe.typeSymbol.primaryConstructor.info.paramTypes match {
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/LambdaLift.scala:450]: The indentation looks off here...

**Comment by soc** [comment]
LGTM, although `BooleanRef zero()` instead of `BooleanRef true()` and `BooleanRef false()` ... well, I guess consistency with the other types wins this time.

**Comment by magarciaEPFL** [comment]
thanks for the review, @soc !

**Comment by paulp** [comment]
I don't know all the background of this but I support this patch on its own merits. LGTM.

**Pull request merged**